host/xtest/pkcs11_1000.c

```diff
@@ -3741,3 +3741,430 @@ static void xtest_pkcs11_test_1016(ADBG_Case_t *c)
+			    CKK_AES, CK_FALSE, CK_FALSE);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_GetAttributeValue(session, derv_key_hdl, get_template,
+				 ARRAY_SIZE(get_template) - 1);
+	if (!ADBG_EXPECT_CK_OK(c, rv) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, secret_len, ==, key_len) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_class, ==, CKO_SECRET_KEY) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_key_type, ==, CKK_AES) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_local, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_sensitive, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_extract, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_asensitive, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_nextract, ==, CK_TRUE))
+		goto out;
+
+	rv = C_DestroyObject(session, derv_key_hdl);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Derive Generic secret key using AES-CBC");
+	/*
+	 * Use AES key 2 as Parent key
+	 * 1. VALUE_LEN < DATA_LEN, Derived key should have VALUE_LEN key size
+	 * 2. Derivation template has SENSITIVE = FALSE, EXTRACTABLE = TRUE
+	 * Parent key has ALWAYS_SENSITIVE = TRUE, NEVER_EXTRACTABLE = TRUE
+	 * So derived key, ALWAYS_SENSITIVE will be same as SENSITIVE and
+	 * NEVER_EXTRACTABLE will be opposite of EXTRACTABLE
+	 * 3. LOCAL should be false
+	 */
+	data_len = 512;
+	key_len = 256;
+	rv = derive_sym_key(session, aes_key2, CKM_AES_CBC_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_GENERIC_SECRET, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	rv = C_GetAttributeValue(session, derv_key_hdl, get_template,
+				 ARRAY_SIZE(get_template));
+	if (!ADBG_EXPECT_CK_OK(c, rv) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, get_template[idx].ulValueLen, ==,
+					  key_len) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, secret_len, ==, key_len) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_class, ==, CKO_SECRET_KEY) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_key_type, ==,
+					  CKK_GENERIC_SECRET) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_local, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_sensitive, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_extract, ==, CK_TRUE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_asensitive, ==, CK_FALSE) ||
+	    !ADBG_EXPECT_COMPARE_UNSIGNED(c, g_nextract, ==, CK_FALSE))
+		goto out;
+
+	rv = C_DestroyObject(session, derv_key_hdl);
+	if (!ADBG_EXPECT_CK_OK(c, rv))
+		goto out;
+
+	Do_ADBG_EndSubCase(c, NULL);
+
+	Do_ADBG_BeginSubCase(c, "Invalid parameters during derivation");
+
+	/* Length of data used for derivation < key length */
+	data_len = 16;
+	key_len = 32;
+	rv = derive_sym_key(session, aes_key1, CKM_AES_ECB_ENCRYPT_DATA,
+			    data_len, &derv_key_hdl, key_len, CKO_SECRET_KEY,
+			    CKK_AES, CK_FALSE, CK_TRUE);
+	if (!ADBG_EXPECT_CK_RESULT(c, CKR_DATA_LEN_RANGE, rv))
```

Comment:
Return value not in list of expected return values for `C_DeriveKey()`.

Perhaps `CKR_ARGUMENTS_BAD`.

Reply:
Section 2.15.3 in "PKCS #11 Cryptographic Token Interface Current Mechanisms Specification Version 2.40 Plus Errata 01" mentions this error for these mechanisms. I will probably extend the list in optee_client to add this error.

Reply:
Ok. I think the adjustment in libckteec is the right approach in this case.

Perhaps you could carry a note in commit message and in the code why there is extra return value.

What I do is use my favorite comparison program and paste both code and specification and then just compare that they match -- in this case there would be extra value and would instantly think that it is incorrect.

Reply:
Added in optee_client